File: rmt-services/RMT.Identity/src/common/decorators/user.decorator.ts
import { createParamDecorator, ExecutionContext } from '@nestjs/common';

export interface IUser {
  email: string;
  name: string;
  designation: string;
  service_line: string;
  emp_code: string;
  roles: string[];
  token: string;
  id: number;
  role: string;
  supercoach_name: string;
  co_supercoach_name: string;
  employee_id: string;
  uemail_id: string;
  competency: string;
  competencyId: string;
  app_permissions: any[];
  access_token?: string;
}

export const User = createParamDecorator(
  (data: unknown, ctx: ExecutionContext) => {
    const request = ctx.switchToHttp().getRequest();

    if (!request?.headers?.userinfo) {
      return {
        email: '',
        name: '',
        service_line: '',
        designation: '',
        roles: [],
        emp_code: '',
        token: '',
        role: '',
        supercoach_name: '',
        co_supercoach_name: '',
        employee_id: '',
        uemail_id: '',
        competency: '',
        competencyId: '',
        app_permissions: [],
      };
    }
    const {
      name,
      email,
      designation,
      service_line,
      roles,
      emp_code,
      role,
      supercoach_name,
      co_supercoach_name,
      employee_id,
      uemail_id,
      competency,
      competencyId,
      app_permissions,
      token,
    } = JSON.parse(request?.headers?.userinfo);

    return {
      email,
      name,
      service_line,
      designation,
      roles,
      emp_code,
      role,
      supercoach_name,
      co_supercoach_name,
      employee_id,
      uemail_id,
      competency,
      competencyId,
      app_permissions,
      token: request.headers.authorization,
    };
  },
);

File: rmt-services/RMT.Configuration/RMT.Configuration.Domain/UserDecorator.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace RMT.Configuration.Domain
{
    /// <summary>
    /// UserDecorator
    /// </summary>
    public class UserDecorator
    {
        public int? id { get; set; }
        public string? email { get; set; }
        public string? name { get; set; }
        public string? emp_code { get; set; }
        public string? designation { get; set; }
        public string? service_line { get; set; }
        public string[]? roles { get; set; }
        public string? role { get; set; }
        public string? supercoach_name { get; set; }
        public string? co_supercoach_name { get; set; }
        public string? supercoach_mid { get; set; }
        public string? co_supercoach_mid { get; set; }
        public string? employee_id { get; set; }
        public string? uemail_id { get; set; }
        public string? competency { get; set; }
        public string? competencyId { get; set; }
        public string? business_unit { get; set; }
        public string? location { get; set; }
        public string? grade { get; set; }
        public List<ModulePermissionDTO>? app_permissions { get; set; }
    }

    /// <summary>
    /// ModulePermissionDTO
    /// </summary>
    public class ModulePermissionDTO
    {
        public string module_name { get; set; }
        public int module_id { get; set; }
        public PermissionLevel permissions { get; set; }
        public bool is_assigned { get; set; }
    }

    /// <summary>
    /// Permission
    /// </summary>
    public class PermissionLevel
    {
        public bool read { get; set; }
        public bool update { get; set; }
        public bool create { get; set; }
        public bool delete { get; set; }
        public bool approve { get; set; }
    }
}
